### ISSUE TITLE
React Server Components (RSC) support for Contentlayer

### ISSUE THREAD

**Issue by schickling** [contentlayerdev/contentlayer#311]
[React Server Components](https://nextjs.org/docs/advanced-features/react-18/server-components) are about to become more generally available. We should investigate what's the best way to use Contentlayer in a RSC setup.

Status: Currently blocked by https://github.com/vercel/next.js/issues/41865

**Comment by heikir** [comment]
Any updates on this since Next13 was publicly launched yesterday?

It seems the setup currently throws with an error:

`Error: Module not found: Package path ./client is not exported from package /Users/heiki/Development/next13/.yarn/unplugged/contentlayer-npm-0.2.8-c370ea5865/node_modules/contentlayer/ (see exports field in /Users/heiki/Development/next13/.yarn/unplugged/contentlayer-npm-0.2.8-c370ea5865/node_modules/contentlayer/package.json)
>  1 | // NOTE This file is auto-generated by Contentlayer
  2 | 
  3 | export { isType } from 'contentlayer/client'
  4 | 
  5 | // NOTE During development Contentlayer imports from `.mjs` files to improve HMR speeds.
  6 | // During (production) builds Contentlayer it imports from `.json` files to improve build performance.`

**Comment by schickling** [comment]
Update: I've started the work of exploring how Contentlayer would be be used with Next.js 13 and RSC. Unfortunately this work is currently blocked by a bug in Next.js (https://github.com/vercel/next.js/issues/41865) however it seems like @balazsorban44 is working on fixing it.

**Comment by heikir** [comment]
Thanks for the update, @schickling!

**Comment by shamwela** [comment]
https://github.com/vercel/next.js/issues/41865 It's fixed now.

**Comment by ggwilin** [comment]
I updated my next to `v13.0.4` and it works!

**Comment by schickling** [comment]
Update: Seems like RSC is ready to be used - we just need to update our documentation (cc @seancdavis). In the meanwhile you can already refer to [this example](https://github.com/contentlayerdev/next-contentlayer-example/tree/experiment/rsc).

**Comment by yeroshek** [comment]
Thank you, @schickling! Is there an example for MDX? I’m using `useMDXComponent` from `next-contentlayer/hooks`, and seems to not work with RSC

**Comment by seancdavis** [comment]
@yeroshek Can you share more information about what isn't working? What version are you using and what error(s) are you seeing?

**Comment by simonporter007** [comment]
Is there any example of importing client components? I was trying to add a component using a useState hook with the `useMDXComponent()`, passed the imported component, which has the 'use client' directive:

```
const Component = useMDXComponent(post?.body?.code || '')
...
<article className="prose-dark mx-auto max-w-3xl ">
    <Component components={{ HeaderLink }} />
</article>
```

and it seems to be throwing the error:

```
[0] error - node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/build/webpack/loaders/next-flight-loader/module-proxy.js (56:14) @ Object.get
[0] error - Error: Cannot access HeaderLink.propTypes on the server. You cannot dot into a client module from a server component. You can only pass the imported name through.
[0]     at Object.get (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/build/webpack/loaders/next-flight-loader/module-proxy.js:56:15)
[0]     at ar (eval at getMDXComponent (webpack-internal:///(sc_server)/./node_modules/.pnpm/next-contentlayer@0.3.0_mbdnv6f2bnj44jkb6khzbtv73u/node_modules/next-contentlayer/dist/hooks/useMDXComponent.js), <anonymous>:13:823)
[0]     at ir (eval at getMDXComponent (webpack-internal:///(sc_server)/./node_modules/.pnpm/next-contentlayer@0.3.0_mbdnv6f2bnj44jkb6khzbtv73u/node_modules/next-contentlayer/dist/hooks/useMDXComponent.js), <anonymous>:13:2637)
[0]     at Te (eval at getMDXComponent (webpack-internal:///(sc_server)/./node_modules/.pnpm/next-contentlayer@0.3.0_mbdnv6f2bnj44jkb6khzbtv73u/node_modules/next-contentlayer/dist/hooks/useMDXComponent.js), <anonymous>:13:3082)
[0]     at _r (eval at getMDXComponent (webpack-internal:///(sc_server)/./node_modules/.pnpm/next-contentlayer@0.3.0_mbdnv6f2bnj44jkb6khzbtv73u/node_modules/next-contentlayer/dist/hooks/useMDXComponent.js), <anonymous>:54:815)
[0]     at attemptResolveElement (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1276:42)
[0]     at resolveModelToJSON (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1745:53)
[0]     at Object.toJSON (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1133:40)
[0]     at stringify (<anonymous>)
[0]     at processModelChunk (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:172:36)
[0]     at retryTask (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1960:50)
[0]     at performWork (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1998:33)
[0]     at eval (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1373:40)
[0]     at scheduleWork (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:52:25)
[0]     at pingTask (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1372:29)
[0]     at ping (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1385:40)
[0]     at processTicksAndRejections (node:internal/process/task_queues:96:5) {
[0]   page: '/posts/post-01'
[0] }
```
just wondered if this was what others were seeing?

**Comment by juliusmarminge** [comment]
Now that Next.js supports mdx as server components, can we get support / an example / docs for how to use contentlayer in rsc?

Ref Next.js docs: https://beta.nextjs.org/docs/guides/mdx

**Comment by wootsbot** [comment]
> Is there any example of importing client components? I was trying to add a component using a useState hook with the `useMDXComponent()`, passed the imported component, which has the 'use client' directive:
> 
> ```
> const Component = useMDXComponent(post?.body?.code || '')
> ...
> <article className="prose-dark mx-auto max-w-3xl ">
>     <Component components={{ HeaderLink }} />
> </article>
> ```
> 
> and it seems to be throwing the error:
> 
> ```
> [0] error - node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/build/webpack/loaders/next-flight-loader/module-proxy.js (56:14) @ Object.get
> [0] error - Error: Cannot access HeaderLink.propTypes on the server. You cannot dot into a client module from a server component. You can only pass the imported name through.
> [0]     at Object.get (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/build/webpack/loaders/next-flight-loader/module-proxy.js:56:15)
> [0]     at ar (eval at getMDXComponent (webpack-internal:///(sc_server)/./node_modules/.pnpm/next-contentlayer@0.3.0_mbdnv6f2bnj44jkb6khzbtv73u/node_modules/next-contentlayer/dist/hooks/useMDXComponent.js), <anonymous>:13:823)
> [0]     at ir (eval at getMDXComponent (webpack-internal:///(sc_server)/./node_modules/.pnpm/next-contentlayer@0.3.0_mbdnv6f2bnj44jkb6khzbtv73u/node_modules/next-contentlayer/dist/hooks/useMDXComponent.js), <anonymous>:13:2637)
> [0]     at Te (eval at getMDXComponent (webpack-internal:///(sc_server)/./node_modules/.pnpm/next-contentlayer@0.3.0_mbdnv6f2bnj44jkb6khzbtv73u/node_modules/next-contentlayer/dist/hooks/useMDXComponent.js), <anonymous>:13:3082)
> [0]     at _r (eval at getMDXComponent (webpack-internal:///(sc_server)/./node_modules/.pnpm/next-contentlayer@0.3.0_mbdnv6f2bnj44jkb6khzbtv73u/node_modules/next-contentlayer/dist/hooks/useMDXComponent.js), <anonymous>:54:815)
> [0]     at attemptResolveElement (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1276:42)
> [0]     at resolveModelToJSON (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1745:53)
> [0]     at Object.toJSON (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1133:40)
> [0]     at stringify (<anonymous>)
> [0]     at processModelChunk (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:172:36)
> [0]     at retryTask (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1960:50)
> [0]     at performWork (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1998:33)
> [0]     at eval (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1373:40)
> [0]     at scheduleWork (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:52:25)
> [0]     at pingTask (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1372:29)
> [0]     at ping (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1385:40)
> [0]     at processTicksAndRejections (node:internal/process/task_queues:96:5) {
> [0]   page: '/posts/post-01'
> [0] }
> ```
> 
> just wondered if this was what others were seeing?

Have you found how to introduce a client component?

**Comment by simonporter007** [comment]
> > Is there any example of importing client components? I was trying to add a component using a useState hook with the `useMDXComponent()`, passed the imported component, which has the 'use client' directive:
> > ```
> > const Component = useMDXComponent(post?.body?.code || '')
> > ...
> > <article className="prose-dark mx-auto max-w-3xl ">
> >     <Component components={{ HeaderLink }} />
> > </article>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and it seems to be throwing the error:
> > ```
> > [0] error - node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/build/webpack/loaders/next-flight-loader/module-proxy.js (56:14) @ Object.get
> > [0] error - Error: Cannot access HeaderLink.propTypes on the server. You cannot dot into a client module from a server component. You can only pass the imported name through.
> > [0]     at Object.get (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/build/webpack/loaders/next-flight-loader/module-proxy.js:56:15)
> > [0]     at ar (eval at getMDXComponent (webpack-internal:///(sc_server)/./node_modules/.pnpm/next-contentlayer@0.3.0_mbdnv6f2bnj44jkb6khzbtv73u/node_modules/next-contentlayer/dist/hooks/useMDXComponent.js), <anonymous>:13:823)
> > [0]     at ir (eval at getMDXComponent (webpack-internal:///(sc_server)/./node_modules/.pnpm/next-contentlayer@0.3.0_mbdnv6f2bnj44jkb6khzbtv73u/node_modules/next-contentlayer/dist/hooks/useMDXComponent.js), <anonymous>:13:2637)
> > [0]     at Te (eval at getMDXComponent (webpack-internal:///(sc_server)/./node_modules/.pnpm/next-contentlayer@0.3.0_mbdnv6f2bnj44jkb6khzbtv73u/node_modules/next-contentlayer/dist/hooks/useMDXComponent.js), <anonymous>:13:3082)
> > [0]     at _r (eval at getMDXComponent (webpack-internal:///(sc_server)/./node_modules/.pnpm/next-contentlayer@0.3.0_mbdnv6f2bnj44jkb6khzbtv73u/node_modules/next-contentlayer/dist/hooks/useMDXComponent.js), <anonymous>:54:815)
> > [0]     at attemptResolveElement (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1276:42)
> > [0]     at resolveModelToJSON (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1745:53)
> > [0]     at Object.toJSON (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1133:40)
> > [0]     at stringify (<anonymous>)
> > [0]     at processModelChunk (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:172:36)
> > [0]     at retryTask (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1960:50)
> > [0]     at performWork (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1998:33)
> > [0]     at eval (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1373:40)
> > [0]     at scheduleWork (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:52:25)
> > [0]     at pingTask (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1372:29)
> > [0]     at ping (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.1.7-canary.21_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1385:40)
> > [0]     at processTicksAndRejections (node:internal/process/task_queues:96:5) {
> > [0]   page: '/posts/post-01'
> > [0] }
> > ```

> > just wondered if this was what others were seeing?
> 
> Have you found how to introduce a client component?

@wootsbot I had to extract the `useMdxComponent` call to its own component, and then use the `"use client"` directive at the top of that file. From what I can see the component that calls useMdxComponent needs to be a client component when passing any component that is _also_ a client component. However this makes all components passed to the hook also client components too. 😕 

Though it works fine as a server component if you don't pass any client components with the `compoennts={}`. The way I'm working with it for now is to have separate component which uses `useMdxComponent`, one for the server version and one for the client version. If I don't use client components in the MDX, then I can render the RSC version, otherwise, it gets all client components instead.

**Comment by sporter1-godaddy** [comment]
Thanks for the new release @schickling ! Just a note, as I don't think it warrants a new issue as such, just that from the release notes I see you mention:

> Note: While it's theoretically also possible to use Contentlayer combined with the 'use client' approach, we don't recommend it as it massively increases page sizes and thus the page load time.

but the full example with next RSC's uses plain markdown files, no custom components/jsx. If possible, when updating the docs, it would be great to see how to avoid 'use client' yet still use client components in the mdx or have a more involved sample provided than just plain markdown. Plain markdown works a treat of course!

From my testing, I haven't found a way around that because I think you would need to pass the client component as a child e.g.

```
<Component components={serverComponents}>
    <ClientComponent />
</Component>
```
but the only way around that I found was to make the component using useMdxComponent a client component, and then passing the client components to it normally, which as you say, isn't ideal.

**Comment by sporter1-godaddy** [comment]
@juliusmarminge don't want to perpetuate a closed thread, but as we're both here :sweat_smile: 

I am using a format closer to your second example actually, so just thought I'd share.

```
posts/
  post1.mdx
  post2.mdx
  ...
app/
  posts/
    [...slug]
      page.tsx // parse slug and read from content, this is for posts with server components only
    post2
      page.tsx // static routes take precedence over dynamic above, means I can use this post as a client component
contentlayer.config.js // this defines my Post type and fields, and mdx plugins
next.config.mjs
```

where the contentlayer config searches for my mdx files in the posts dir:
```
export default makeSource({
  contentDirPath: 'posts',
  documentTypes: [Post],
  mdx: {
    remarkPlugins: [remarkGfm, remarkGemoji],
    rehypePlugins: [
      rehypeMeta,
      rehypeInferReadingTimeMeta,
      rehypeSlug,
      rehypeAutolinkHeadings,
      rehypePrism,
      rehypeAccessibleEmojis,
    ],
  },
})
```

but ya, would definitely like to see an easier way to pass child components into the tree without needing to make the useMdxComponent a client component too, then I could do away with the individual post dir's and just rely on `[slug]`
